File: firefly-example/src/test/java/com/firefly/example/test/dao/TestInventoryDAO.java
package com.firefly.example.test.dao;

import com.firefly.example.reactive.coffee.store.dao.InventoryDAO;
import com.firefly.example.reactive.coffee.store.model.Inventory;
import com.firefly.example.reactive.coffee.store.vo.InventoryOperator;
import com.firefly.example.reactive.coffee.store.vo.InventoryUpdate;
import com.firefly.example.test.TestBase;
import com.firefly.reactive.adapter.db.ReactiveSQLClient;
import org.junit.Assert;
import org.junit.Test;
import reactor.test.StepVerifier;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static com.firefly.example.reactive.coffee.store.AppMain.javaCtx;
import static org.hamcrest.Matchers.is;

/**
 * @author Pengtao Qiu
 */
public class TestInventoryDAO extends TestBase {

    private ReactiveSQLClient db = javaCtx.getBean(ReactiveSQLClient.class);
    private InventoryDAO inventoryDAO = javaCtx.getBean(InventoryDAO.class);

    @Test
    public void testUpdate() {
        List<InventoryUpdate> list = new ArrayList<>();
        InventoryUpdate update = new InventoryUpdate();
        update.setAmount(10L);
        update.setProductId(4L);
        list.add(update);

        update = new InventoryUpdate();
        update.setAmount(20L);
        update.setProductId(5L);
        list.add(update);

        List<InventoryUpdate> update1 = list;
        StepVerifier.create(db.newTransaction(c -> inventoryDAO.updateBatch(update1, InventoryOperator.SUB, c))).assertNext(r -> {
            Assert.assertThat(r.length, is(2));
            Arrays.stream(r).forEach(row -> Assert.assertThat(row, is(1)));
        }).expectComplete().verify();

        StepVerifier.create(inventoryDAO.listByProductId(Arrays.asList(4L, 5L))).assertNext(r -> {
            Map<Long, Inventory> map = r.parallelStream().collect(Collectors.toMap(Inventory::getProductId, v -> v));
            Assert.assertThat(map.get(4L).getAmount(), is(67L));
            Assert.assertThat(map.get(5L).getAmount(), is(80L));
        }).expectComplete().verify();

        list = new ArrayList<>();
        update = new InventoryUpdate();
        update.setAmount(90L);
        update.setProductId(4L);
        list.add(update);

        update = new InventoryUpdate();
        update.setAmount(20L);
        update.setProductId(5L);
        list.add(update);

        List<InventoryUpdate> update2 = list;
        StepVerifier.create(db.newTransaction(c -> inventoryDAO.updateBatch(update2, InventoryOperator.SUB, c))).assertNext(r -> {
            Assert.assertThat(r.length, is(2));
            Assert.assertThat(r[0], is(0));
            Assert.assertThat(r[1], is(1));
        }).expectComplete().verify();

        StepVerifier.create(inventoryDAO.listByProductId(Arrays.asList(4L, 5L))).assertNext(r -> {
            Map<Long, Inventory> map = r.parallelStream().collect(Collectors.toMap(Inventory::getProductId, v -> v));
            Assert.assertThat(map.get(4L).getAmount(), is(67L));
            Assert.assertThat(map.get(5L).getAmount(), is(60L));
        }).expectComplete().verify();
    }
}

File: firefly-example/src/test/kotlin/com/firefly/example/kotlin/test/dao/TestInventoryDAO.kt
package com.firefly.example.kotlin.test.dao

import com.firefly.example.kotlin.coffee.store.dao.InventoryDAO
import com.firefly.example.kotlin.coffee.store.model.Inventory
import com.firefly.example.kotlin.coffee.store.vo.InventoryOperator
import com.firefly.example.kotlin.coffee.store.vo.InventoryUpdate
import com.firefly.example.kotlin.test.TestBase
import com.firefly.kotlin.ext.context.Context
import kotlinx.coroutines.experimental.runBlocking
import org.junit.Test
import java.util.stream.Collectors
import kotlin.test.assertEquals

/**
 * @author Pengtao Qiu
 */
class TestInventoryDAO : TestBase() {

    private val inventoryDAO = Context.getBean<InventoryDAO>()

    @Test
    fun test(): Unit = runBlocking {
        val ret = inventoryDAO.updateBatch(
                listOf(InventoryUpdate(4L, 10L),
                       InventoryUpdate(5L, 20L)),
                InventoryOperator.SUB)
        assertEquals(2, ret.size)
        ret.forEach { assertEquals(it, 1) }

        val map = inventoryDAO.listByProductId(listOf(4L, 5L))
                .parallelStream().collect(Collectors.toMap(Inventory::productId, { v -> v }))
        assertEquals(67L, map[4L]?.amount)
        assertEquals(80L, map[5L]?.amount)

        val ret2 = inventoryDAO.updateBatch(
                listOf(InventoryUpdate(4L, 90L),
                       InventoryUpdate(5L, 20L)),
                InventoryOperator.SUB)
        assertEquals(2, ret2.size)
        assertEquals(0, ret2[0])
        assertEquals(1, ret2[1])

        val map2 = inventoryDAO.listByProductId(listOf(4L, 5L))
                .parallelStream().collect(Collectors.toMap(Inventory::productId, { v -> v }))
        assertEquals(67L, map2[4L]?.amount)
        assertEquals(60L, map2[5L]?.amount)
        Unit
    }
}